tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs

```diff
@@ -0,0 +1,2825 @@
+use crate::integer::keycache::KEY_CACHE;
+use crate::integer::{RadixCiphertext, RadixClientKey, ServerKey};
+use crate::shortint::parameters::*;
+use rand::Rng;
+
+/// Number of loop iteration within randomized tests
+const NB_TEST: usize = 30;
+
+/// Smaller number of loop iteration within randomized test,
+/// meant for test where the function tested is more expensive
+const NB_TEST_SMALLER: usize = 10;
+const NB_CTXT: usize = 4;
+
+/// helper function to do a rotate left when the type used to store
+/// the value is bigger than the actual intended bit size
+fn rotate_left_helper(value: u64, n: u32, actual_bit_size: u32) -> u64 {
+    // We start with:
+    // [0000000000000|xxxx]
+    // 64           b    0
+    //
+    // rotated will be
+    // [0000000000xx|xx00]
+    // 64           b    0
+    let n = n % actual_bit_size;
+    let mask = 1u64.wrapping_shl(actual_bit_size) - 1;
+    let shifted_mask = mask.wrapping_shl(n) & !mask;
+
+    let rotated = value.rotate_left(n);
+
+    (rotated & mask) | ((rotated & shifted_mask) >> actual_bit_size)
+}
+
+/// helper function to do a rotate right when the type used to store
+/// the value is bigger than the actual intended bit size
+fn rotate_right_helper(value: u64, n: u32, actual_bit_size: u32) -> u64 {
+    // We start with:
+    // [0000000000000|xxxx]
+    // 64           b    0
+    //
+    // mask: [000000000000|mmmm]
+    // shifted_ mask: [mm0000000000|0000]
+    //
+    // rotated will be
+    // [xx0000000000|00xx]
+    // 64           b    0
+    //
+    // To get the 'cycled' bits where they should be,
+    // we get them using a mask then shift
+    let n = n % actual_bit_size;
+    let mask = 1u64.wrapping_shl(actual_bit_size) - 1;
+    // shifted mask only needs the bits that cycled
+    let shifted_mask = mask.rotate_right(n) & !mask;
+
+    let rotated = value.rotate_right(n);
+
+    (rotated & mask) | ((rotated & shifted_mask) >> (u64::BITS - actual_bit_size))
+}
+
+/// This trait is to be implemented by a struct that is capable
+/// of executing a particular function to be tested.
+pub(crate) trait FunctionExecutor<TestInput, TestOutput> {
+    /// Setups the executor
+    ///
+    /// Implementors are expected to be fully functional after this
+    /// function has been called.
+    fn setup(&mut self, cks: &RadixClientKey, sks: ServerKey);
```

Comment:
As briefly discussed, the goal of this setup function is for the executor to setup itself so
that it is able to work on ciphertexts that will be encrypted/decrypted by the given cks.

The implemetor can do whatever it wants to achieve that.

If in the future and with evolutions that will happen, if that proves to be  a bit too restrictive, then we would inverse the paradigme. That is the test case would receive both the executor that can execute the function to be tested and a `tfhe::integer::ClientKey` that corresponds to what the executor is capable of working on.